components/eam/src/chemistry/modal_aero/seasalt_model.F90

```diff
@@ -348,48 +343,147 @@ subroutine ocean_data_readnl(nlfile)
+ 
+    call seasalt_emisflx_calc(ncol, fi, ocnfrc, emis_scale, mass_flx_flag, & ! in
+                              cflx)                                          ! inout
+
+end subroutine seasalt_emis
+
+
+subroutine seasalt_emisflx_calc(ncol, fi, ocnfrc, emis_scale, flx_type, & ! in
+                                cflx)                                     ! inout
+   
+    use sslt_sections, only: nsections, fluxes, Dg, rdry
+    use mo_constants,  only: dns_aer_sst=>seasalt_density, pi
+
+    ! input
+    integer, intent(in)  :: ncol
+    real(r8), intent(in) :: fi(pcols, nsections)        ! sea salt number fluxes in each size bin [#/m2/s]
+    real(r8), intent(in) :: ocnfrc(pcols)               ! ocean fraction [unitless]
+    real(r8), intent(in) :: emis_scale                  ! sea salt emission tuning factor [unitless]
+    integer, intent(in)  :: flx_type                    ! number ==0 or mass ==1 flux to be calculated
+
+    ! output
+    real(r8), intent(inout) :: cflx(:,:)      ! mass and number emission fluxes for aerosols [kg/m2/s or #/m2/s]
+
+    ! local vars
+    integer  :: ispec, ibin
+    integer  :: mode_idx
+    real(r8) :: cflx_tmp1(pcols)              ! temp array for calculating emission fluxes [kg/m2/s or #/m2/s]
+
+
+    if (flx_type == 0) then
+
+       do ispec = 1,nslt
+          mode_idx  = seasalt_indices(nslt+nslt_om+ispec)  ! Index of number mode
+      
+          if (mode_idx>0) then
+
+             ! Total number flux per mode
+             do ibin=1, nsections
+                cflx_tmp1(:ncol) = 0.0_r8
+
+                if (Dg(ibin) >= sst_sz_range_lo(ispec) .and. Dg(ibin) < sst_sz_range_hi(ispec)) then
+                   cflx_tmp1(:ncol) = fi(:ncol,ibin)*ocnfrc(:ncol)*emis_scale  !++ ag: scale sea-salt
+                
+                   ! Mixing state 3: internal mixture, add OM to mass and number
+                   cflx(:ncol,mode_idx) = cflx(:ncol,mode_idx) + cflx_tmp1(:ncol)
+                endif
+             enddo
+          endif
+       enddo
+   
+    else
+
+       do ispec = 1,nslt
+          mode_idx = seasalt_indices(ispec)  ! Index of mass mode
+                                                           
+          cflx(:ncol,mode_idx) = 0.0_r8
+
+          ! Total mass flux per mode
+          do ibin=1, nsections
+             if (Dg(ibin) >= sst_sz_range_lo(ispec) .and. Dg(ibin) < sst_sz_range_hi(ispec)) then
+                cflx_tmp1(:ncol) = 0.0_r8
+                cflx_tmp1(:ncol) = fi(:ncol,ibin)*ocnfrc(:ncol)*emis_scale  &   !++ ag: scale sea-salt
+                               *4._r8/3._r8*pi*rdry(ibin)**3*dns_aer_sst  ! should use dry size, convert from number to mass flux (kg/m2/s)
+
+                ! Mixing state 3: internal mixture, add OM to mass and number
+                cflx(:ncol,mode_idx) = cflx(:ncol,mode_idx) + cflx_tmp1(:ncol)
+             endif
+          enddo
+       enddo
+
+    endif
+end subroutine seasalt_emisflx_calc
```

Comment:
Would you please try replacing the code of this subroutine with the following to see if it is stays BFB?
```
integer :: num_idx_append
num_idx_append = 0 !index to append for mass fluxes
    if (flx_type == 0) num_idx_append = nslt+nslt_om ! if mass flux, append appropriate index
    
    do ispec = 1,nslt
       mode_idx  = seasalt_indices(num_idx_append+ispec)  ! Index of number mode
       if (mode_idx>0) then
          if (flx_type == 1) cflx(:ncol,mode_idx) = 0.0_r8 ! for number fluxes, zero-out the fluxes
          ! Total number flux per mode
          do ibin=1, nsections
             cflx_tmp1(:ncol) = 0.0_r8
             if (Dg(ibin) >= sst_sz_range_lo(ispec) .and. Dg(ibin) < sst_sz_range_hi(ispec)) then
                cflx_tmp1(:ncol) = fi(:ncol,ibin)*ocnfrc(:ncol)*emis_scale !++ ag: scale sea-salt

                !For number fluxes, multiply by the diameter
                if (flx_type == 1) cflx_tmp1(:ncol) = cflx_tmp1(:ncol) * 4._r8/3._r8*pi*rdry(ibin)**3*dns_aer_sst
                
                ! Mixing state 3: internal mixture, add OM to mass and number
                cflx(:ncol,mode_idx) = cflx(:ncol,mode_idx) + cflx_tmp1(:ncol)
             endif
          enddo
       endif
    enddo
```

Reply:
Thanks for the suggestions. It works fine. I was thinking to reducing if within the do loops.

Reply:
Yeah, that's a good point. I will let the C++/Kokkos team decide if it is a performance issue. I wanted to keep the repeating codes to a minimal here. As you know some of this code can still be simplified but it breaks BFB.